jaxley/connect.py

```diff
@@ -44,33 +44,52 @@ def fully_connect(
     pre_cell_view: "View",
     post_cell_view: "View",
     synapse_type: "Synapse",
+    random_post_comp: bool = False,
 ):
     """Appends multiple connections which build a fully connected layer.
 
-    Connections are from branch 0 location 0 to a randomly chosen branch and loc.
+    Connections are from branch 0 location 0 of the pre-synaptic cell to branch 0
+    location 0 of the post-synaptic cell unless random_post_comp=True.
 
     Args:
         pre_cell_view: View of the presynaptic cell.
         post_cell_view: View of the postsynaptic cell.
         synapse_type: The synapse to append.
+        random_post_comp: If True, randomly samples the postsynaptic compartments.
     """
     # Get pre- and postsynaptic cell indices.
     num_pre = len(pre_cell_view._cells_in_view)
     num_post = len(post_cell_view._cells_in_view)
 
-    # Infer indices of (random) postsynaptic compartments.
-    global_post_indices = (
-        post_cell_view.nodes.groupby("global_cell_index")
-        .sample(num_pre, replace=True)
-        .index.to_numpy()
-    )
-    global_post_indices = global_post_indices.reshape((-1, num_pre), order="F").ravel()
-    post_rows = post_cell_view.nodes.loc[global_post_indices]
-
-    # Pre-synapse is at the zero-eth branch and zero-eth compartment.
-    pre_rows = pre_cell_view.scope("local").branch(0).comp(0).nodes.copy()
-    # Repeat rows `num_post` times. See SO 50788508.
-    pre_rows = pre_rows.loc[pre_rows.index.repeat(num_post)].reset_index(drop=True)
+    # Get the indices of the connections, like it's a fully connected connectivity matrix
+    from_idx = np.repeat(range(0, num_pre), num_post)
+    to_idx = np.tile(
+        range(0, num_post), num_pre
+    )  # used only if random_post_comp is False
+
+    # Pre-synapse at the zero-eth branch and zero-eth compartment
+    global_pre_comp_indices = (
+        pre_cell_view.scope("local").branch(0).comp(0).nodes.index.to_numpy()
```

Comment:
can also be done with `pre_cell_view.nodes.groupby("global_cell_index").first()["global_comp_index"]`, then you dont have to do all the scoping